From 87aad06746d76c5eb4d09b067fee2ca722f42493 Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Sat, 26 Oct 2024 15:39:50 +0530
Subject: [PATCH 2/10] wifi: ath12k: store sta state in ahsta instead or
 arsta

Currently, the STA state is stored in the arsta structure, which is a
per-link STA structure. However, in mac80211, the STA state is not a
per-link member, the entire STA object has a single state member. If a
link is added when the STA is at state X, it is expected that the link
should also reach state X.

To maintain consistency between mac80211 states and the driver, store the
STA state in the ahsta structure instead of arsta.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c |  2 +-
 drivers/net/wireless/ath/ath12k/core.h |  5 +++--
 drivers/net/wireless/ath/ath12k/mac.c  | 14 ++++++++++----
 3 files changed, 14 insertions(+), 7 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -2512,7 +2512,7 @@ static void ath12k_core_mode1_recovery_s
 		goto key_add;
 	}
 
-	prev_state = arsta->state;
+	prev_state = arsta->ahsta->state;
 	for (state = IEEE80211_STA_NOTEXIST;
 	     state < prev_state; state++) {
 		/* all station set case */
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -814,7 +814,6 @@ struct ath12k_link_sta {
 	u8 sawf_svc_id;
 	/* will be saved to use during recovery */
 	struct ieee80211_key_conf *keys[WMI_MAX_KEY_INDEX + 1];
-	enum ieee80211_sta_state state;
 #ifdef CPTCFG_ATH12K_CFR
 	struct ath12k_per_peer_cfr_capture cfr_capture;
 #endif
@@ -855,7 +854,9 @@ struct ath12k_sta {
 	int ppe_vp_num;
 	struct ath12k_vlan_iface *vlan_iface;
 #endif
- };
+
+	enum ieee80211_sta_state state;
+};
 
 #define ATH12K_HALF_20MHZ_BW 10
 #define ATH12K_2G_MIN_CENTER 2412
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -8514,7 +8514,6 @@ static int ath12k_mac_assign_link_sta(st
 	ahsta->device_bitmap |= BIT(ab->wsi_info.chip_idx);
 	arsta->ahsta = ahsta;
 	arsta->link_id = link_id;
-	arsta->state = IEEE80211_STA_NONE;
 	arsta->is_bridge_peer = is_bridge_peer;
 
 	return 0;
@@ -9105,7 +9104,6 @@ static int ath12k_mac_handle_link_sta_st
 				    arsta->addr);
 	}
 
-	arsta->state = new_state;
 	mutex_unlock(&ar->conf_mutex);
 
 	ath12k_dbg(ar->ab, ATH12K_DBG_PEER, "mac sta %pM old state %d new state :%d\n",
@@ -9388,6 +9386,10 @@ int ath12k_mac_update_sta_state(struct i
 			ret = 0;
 		}
 
+		/* update state if everything went well */
+		if (!ret)
+			ahsta->state = new_state;
+
 		mutex_unlock(&ah->conf_mutex);
 		return ret;
 	}
@@ -9440,7 +9442,7 @@ int ath12k_mac_update_sta_state(struct i
  	}
 
 	/* Reconfig links of arsta during recovery */
-	if (ahsta->link[ahsta->assoc_link_id]->state != IEEE80211_STA_NONE &&
+	if (ahsta->state != IEEE80211_STA_NONE &&
 	    old_state == IEEE80211_STA_NOTEXIST &&
 	    new_state == IEEE80211_STA_NONE) {
 		ahsta->num_peer = 0;
@@ -9529,6 +9531,10 @@ int ath12k_mac_update_sta_state(struct i
 	}
 
 exit:
+	/* update state if everything went well */
+	if (!ret)
+		ahsta->state = new_state;
+
 	mutex_unlock(&ah->conf_mutex);
 
 	return ret;
@@ -16322,7 +16328,7 @@ ath12k_mac_assign_vif_chanctx_handle(str
 
 		mutex_unlock(&ar->conf_mutex);
 		mutex_unlock(&ah->conf_mutex);
-		prev_state = arsta->state;
+		prev_state = arsta->ahsta->state;
 		for (state = IEEE80211_STA_NOTEXIST; state < prev_state;
 		     state++)
 			ath12k_mac_update_sta_state(ar->ah->hw, arvif->ahvif->vif, sta,
